components/board/genericlinux/digital_interrupts.go

```diff
@@ -6,115 +6,159 @@ package genericlinux
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	var err error
+	if len(di.channels) > 0 {
+		err = fmt.Errorf("closed digital interrupt %s, but it still had %d listeners",
+		                 di.config.Name, len(di.channels))
 	}
-	result.startMonitor()
-	return &result, nil
+	return multierr.Combine(err, di.line.Close())
+}
+
+func (di *digitalInterrupt) Name() string {
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	return di.config.Name
 }
 
-func (di *digitalInterrupt) startMonitor() {
-	di.boardWorkers.Add(1)
-	utils.ManagedGo(func() {
-		for {
-			select {
-			case <-di.cancelCtx.Done():
-				return
-			case event := <-di.line.Events():
-				utils.UncheckedError(pinwrappers.Tick(
-					di.cancelCtx, di.interrupt.(*pinwrappers.BasicDigitalInterrupt), event.RisingEdge, uint64(event.Time.UnixNano())))
+func (di *digitalInterrupt) Value(
+	ctx context.Context,
+	extra map[string]interface{},
+) (int64, error) {
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	return di.count, nil
+}
+
+func (di *digitalInterrupt) monitor(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event := <-di.line.Events():
+			di.mu.Lock()
+			defer di.mu.Unlock()
+
+			if event.RisingEdge {
+				di.count++
+			}
+
+			tick := board.Tick{
+				Name: di.config.Name,
+				High: event.RisingEdge,
+				TimestampNanosec: uint64(event.Time.UnixNano()),
+			}
+			for _, ch := range di.channels {
+				select {
+				case <-ctx.Done():
+					return
+				case ch<-tick:
+				}
 			}
 		}
-	}, di.boardWorkers.Done)
+	}
 }
 
-func closeInterrupt(di *digitalInterrupt) error {
-	// We shut down the background goroutine that monitors this interrupt, but don't need to wait
-	// for it to finish shutting down because it doesn't use anything in the line itself (just a
-	// channel of events that the line generates). It will shut down sometime soon, and if that's
-	// after the line is closed, that's fine.
-	di.cancelFunc()
-	return di.line.Close()
+// AddChannel adds a board.Tick channel to stream ticks to.
+func (di *digitalInterrupt) AddChannel(ch chan board.Tick) {
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	di.channels = append(di.channels, ch)
 }
 
-// struct implements board.GPIOPin to support reading current state of digital interrupt pins as GPIO inputs.
-type gpioInterruptWrapperPin struct {
-	interrupt digitalInterrupt
+// RemoveChannel removes a previously-added channel to stream ticks to.
+func (di *digitalInterrupt) RemoveChannel(ch chan board.Tick) {
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	for i, oldCh := range di.channels {
+		if ch != oldCh {
+			continue
+		}
+
+		// To remove this item, move the last item in the list to here and then truncate the list.
+		lastIndex := len(di.channels)-1
+		di.channels[i] = di.channels[lastIndex]
+		di.channels = di.channels[:lastIndex]
+		break
+	}
 }
 
-func (gp gpioInterruptWrapperPin) Set(
+// You can also get the current value from a digitalInterrupt pin. To do this, we provide the
```

Comment:
Im confused why you need the whole gpioPIn interface to use value...cant you just use the digitalinterrupt interface which has name and value?

Reply:
To get the value, wouldn't you call `board.GpioPinByName(...).Get()`? We need something we can return from `GpioPinByName`.

Reply:
oh... I might have misunderstood your question! We could simply implement the `board.DigitalInterrupt` interface and have `.Name()` and `.Value()`, but the old code let you also treat interrupts as read-only GPIO pins. If we no longer supported that, we could get rid of all these GPIO-related functions. ...and maybe we should? I'm unsure, and tried to leave the behavior unchanged. but I'm open to changing it if that's what folks want!

Reply:
oh okay sorry I didnt look closely enough at what this function is doing, its getting the digital value not the count value. I think its helpful to be able to read the current digital value from the digital interrupt so we should keep imo. thanks for explaining!

Reply:
I should put in a comment explaining that better. Thanks for speaking up!